builtin/credential/approle/secret_id.go

```diff
@@ -0,0 +1,500 @@
+
+	// If SecretIDTTL is not specified or if it crosses the backend mount's limit,
+	// cap the expiration to backend's max. Otherwise, use it to determine the
+	// expiration time.
+	if secretEntry.SecretIDTTL < time.Duration(0) || secretEntry.SecretIDTTL > b.System().MaxLeaseTTL() {
+		secretEntry.ExpirationTime = currentTime.Add(b.System().MaxLeaseTTL())
+	} else if secretEntry.SecretIDTTL != time.Duration(0) {
+		// Set the ExpirationTime only if SecretIDTTL was set. SecretIDs should not
+		// expire by default.
+		secretEntry.ExpirationTime = currentTime.Add(secretEntry.SecretIDTTL)
+	}
+
+	// Before storing the SecretID, store its accessor.
+	if err := b.createAccessor(s, secretEntry, secretIDHMAC); err != nil {
+		return nil, err
+	}
+
+	if entry, err := logical.StorageEntryJSON(entryIndex, secretEntry); err != nil {
+		return nil, err
+	} else if err = s.Put(entry); err != nil {
+		return nil, err
+	}
+
+	return secretEntry, nil
+}
+
+// roleIDEntry is used to read the storage entry that maps the
+// RoleID to an Role. This method should be called when the lock
+// for the corresponding SecretID is held.
+func (b *backend) secretIDAccessorEntry(s logical.Storage, secretIDAccessor string) (*secretIDAccessorStorageEntry, error) {
+	if secretIDAccessor == "" {
+		return nil, fmt.Errorf("missing secretIDAccessor")
+	}
+
+	var result secretIDAccessorStorageEntry
+
+	// Create index entry, mroleing the accessor to the token ID
+	entryIndex := "accessor/" + b.salt.SaltID(secretIDAccessor)
+
+	if entry, err := s.Get(entryIndex); err != nil {
+		return nil, err
+	} else if entry == nil {
+		return nil, nil
+	} else if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+// createAccessor creates an identifier for the SecretID. A storage index,
+// mroleing the accessor to the SecretID is also created. This method should
+// be called when the lock for the corresponding SecretID is held.
+func (b *backend) createAccessor(s logical.Storage, entry *secretIDStorageEntry, secretIDHMAC string) error {
+	// Create a random accessor
+	accessorUUID, err := uuid.GenerateUUID()
+	if err != nil {
+		return err
+	}
+	entry.SecretIDAccessor = accessorUUID
+
+	// Create index entry, mroleing the accessor to the token ID
+	entryIndex := "accessor/" + b.salt.SaltID(entry.SecretIDAccessor)
+	if entry, err := logical.StorageEntryJSON(entryIndex, &secretIDAccessorStorageEntry{
+		SecretIDHMAC: secretIDHMAC,
+	}); err != nil {
+		return err
+	} else if err = s.Put(entry); err != nil {
+		return fmt.Errorf("failed to persist accessor index entry: %s", err)
+	}
+
+	return nil
+}
+
+// Iterates through all the Roles, fetches the polices from each Role
+// and returns a union of all the policies combined together. An error
+// is thrown if any Role in the list of Roles supplied is non-existent
+// with the backend.
+func (b *backend) fetchPolicies(s logical.Storage, roles []string) ([]string, error) {
```

Comment:
I think this is left over from AppGroup? I don't see it used anywhere and it doesn't seem like something needed anymore.